polygeist/tools/cgeist/Lib/CGExpr.cc

```diff
@@ -2822,13 +2806,23 @@ ValueCategory MLIRScanner::EmitPointerArithmetic(const BinOpInfo &Info) {
   //   The pointer type is not byte-sized.
   //
   if (BinaryOperator::isNullPointerArithmeticExtension(
-          CGM.getContext(), Opcode, PointerOperand, IndexOperand))
+          CGM.getContext(), Opcode, PointerOperand, IndexOperand)) {
+    if (auto MT = PtrTy.dyn_cast<mlir::MemRefType>()) {
+      auto Val = Builder.create<polygeist::Memref2PointerOp>(
+          Loc,
+          LLVM::LLVMPointerType::get(MT.getElementType(),
+                                     MT.getMemorySpaceAsInt()),
+          Pointer.val);
+      PtrTy = Val.getType();
+    }
+
     return Index.IntToPtr(Builder, Loc, PtrTy);
+  }
```

Comment:
`Val`'s cast is not needed, as we do not need its value. Also, if the original type is a `MemRef`, we would need to cast back after this conversion (we have a `Ptr2MemRef` member function in `ValueCategory`).

It would be nice to have a test for this. I guess copying the current one and changing the result type from `void *` to `char *` would trigger this.

Reply:
After merging the code has changed and now we call `EmitIntegralToPointerConversion`.

Reply:
> It would be nice to have a test for this. I guess copying the current one and changing the result type from `void *` to `char *` would trigger this.

LGTM. If we can add this test, that'd be perfect.